File: BeerViewer/Models/Fleet.cs
using System;
using System.ComponentModel;
using System.Collections.Generic;
using System.Linq;

using BeerViewer.Network;
using BeerViewer.Models.Enums;
using BeerViewer.Models.Raw;
using BeerViewer.Models.kcsapi;
using BeerViewer.Models.Wrapper;

namespace BeerViewer.Models
{
	public class Fleet : TimerNotifier, IIdentifiable
	{
		private readonly Homeport homeport;
		private Ship[] originalShips;

		#region Id Property
		private int _Id;
		public int Id
		{
			get { return this._Id; }
			private set
			{
				if (this._Id != value)
				{
					this._Id = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		#region Name Property
		private string _Name;
		public string Name
		{
			get { return this._Name; }
			internal set
			{
				if (this._Name != value)
				{
					this._Name = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		#region Ships Property
		private Ship[] _Ships = new Ship[0];
		public Ship[] Ships
		{
			get { return this._Ships; }
			private set
			{
				if (this._Ships != value)
				{
					this._Ships = value;
					this.RaisePropertyChanged();
				}
			}
		}

		[EditorBrowsable(EditorBrowsableState.Never)]
		public object ShipsUpdated { get; set; }
		#endregion

		#region FleetSpeed Property
		public ShipSpeed Speed
			=> this.Ships?.Select(x => x.Speed).Min()
				?? ShipSpeed.Immovable;
		#endregion

		#region IsInSortie Property
		private bool _IsInSortie;
		public bool IsInSortie
		{
			get { return this._IsInSortie; }
			set
			{
				if (this._IsInSortie != value)
				{
					this._IsInSortie = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		#region State Property
		private FleetSituation _State { get; set; }
		public FleetSituation State
		{
			get { return this._State; }
			set
			{
				if(this._State != value)
				{
					this._State = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		public Expedition Expedition { get; }

		#region AirSuperiorityPotentialMinimum Property
		private int _AirSuperiorityPotentialMinimum { get; set; }
		public int AirSuperiorityPotentialMinimum
		{
			get { return this._AirSuperiorityPotentialMinimum; }
			private set
			{
				if(this._AirSuperiorityPotentialMinimum != value)
				{
					this._AirSuperiorityPotentialMinimum = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		#region AirSuperiorityPotentialMaximum Property
		private int _AirSuperiorityPotentialMaximum { get; set; }
		public int AirSuperiorityPotentialMaximum
		{
			get { return this._AirSuperiorityPotentialMaximum; }
			private set
			{
				if (this._AirSuperiorityPotentialMaximum != value)
				{
					this._AirSuperiorityPotentialMaximum = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		#region LOS Property
		private double _LOS { get; set; }
		public double LOS
		{
			get { return this._LOS; }
			private set
			{
				if (this._LOS != value)
				{
					this._LOS = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		private int MinimumCondition { get; set; }
		private DateTime? RejuvenateTime { get; set; }

		public bool IsRejuvenating => this.RejuvenateTime.HasValue;

		public TimeSpan? RejuvenateRemaining
			=> !this.IsRejuvenating ? (TimeSpan?)null
			: this.RejuvenateTime.Value < DateTime.Now ? TimeSpan.Zero
			: this.RejuvenateTime.Value - DateTime.Now;

		public string RejuvenateText => this.RejuvenateRemaining.HasValue
			? $"{(int)this.RejuvenateRemaining.Value.TotalHours:D2}:{this.RejuvenateRemaining.Value.ToString(@"mm\:ss")}"
			: "--:--:--";

		internal Fleet(Homeport parent, kcsapi_deck Data)
		{
			this.homeport = parent;
			this.CompositeDisposable.Add(this.Expedition = new Expedition(this));

			this.Update(Data);
		}

		internal void Update(kcsapi_deck Data)
		{
			this.Id = Data.api_id;
			this.Name = Data.api_name;

			this.Expedition.Update(Data.api_mission);
			this.UpdateShips(Data.api_ship.Select(id => this.homeport.Organization.Ships[id]).ToArray());

			this.UpdateCondition();
		}

		#region Sortie, Homing
		internal void Sortie()
		{
			if (!this.IsInSortie)
			{
				this.IsInSortie = true;
				this.UpdateState();
			}
		}
		internal void Homing()
		{
			if (this.IsInSortie)
			{
				this.IsInSortie = false;
				this.UpdateState();
			}
		}
		#endregion

		#region Change, Unset
		internal Ship Change(int index, Ship ship)
		{
			var current = this.originalShips[index];

			List<Ship> list;
			if (index == -1)
			{
				list = this.originalShips.Take(1).ToList();
			}
			else
			{
				list = this.originalShips.ToList();
				list[index] = ship;
				list.RemoveAll(x => x == null);
			}

			var ships = new Ship[this.originalShips.Length];
			Array.Copy(list.ToArray(), ships, list.Count);

			this.UpdateShips(ships);

			return current;
		}

		internal void Unset(int index)
		{
			var list = this.originalShips.ToList();
			list[index] = null;
			list.RemoveAll(x => x == null);

			var ships = new Ship[this.originalShips.Length];
			Array.Copy(list.ToArray(), ships, list.Count);

			this.UpdateShips(ships);
		}

		internal void UnsetAll()
		{
			var list = this.originalShips.Take(1).ToList();
			var ships = new Ship[this.originalShips.Length];
			Array.Copy(list.ToArray(), ships, list.Count);

			this.UpdateShips(ships);
		}

		#endregion

		private void UpdateShips(Ship[] ships)
		{
			this.originalShips = ships;
			this.Ships = ships.Where(x => x != null).ToArray();

			this.CalculateValues();
			this.UpdateState();
		}
		private void UpdateCondition()
		{
			var condition = this.Ships.Min(x => x.Condition);
			var goal = 49;

			// Require time not changed
			if (MinimumCondition == condition)
				return;

			MinimumCondition = condition;
			if (condition >= goal)
			{
				this.RejuvenateTime = (DateTime?)null;
				this.RaisePropertyChanged(nameof(this.IsRejuvenating));
			}
			else
			{
				var rejuvenate = DateTime.Now;

				var value = (goal - condition + 2) / 3 * 3; // Integral dividing
				rejuvenate = rejuvenate.AddMinutes(value);

				this.RejuvenateTime = rejuvenate;
				this.RaisePropertyChanged(nameof(this.IsRejuvenating));
			}

			this.UpdateState();
		}

		private void UpdateState()
		{
			var state = FleetSituation.Empty;

			var ships = this.Ships.ToArray();
			if (ships.Length == 0)
			{
			}
			else {
				if (this.IsInSortie)
					state |= FleetSituation.Sortie;
				else if (this.Expedition.IsInExecution)
					state |= FleetSituation.Expedition;
				else
					state |= FleetSituation.Homeport;
			}

			if (state.HasFlag(FleetSituation.Homeport))
			{
				var repairing = ships.Any(x => this.homeport.Repairyard.CheckRepairing(x.Id));
				if (repairing)
					state |= FleetSituation.Repairing;

				var inShortSupply = ships.Any(s => s.Fuel.Current < s.Fuel.Maximum || s.Ammo.Current < s.Ammo.Maximum);
				if (inShortSupply)
					state |= FleetSituation.InShortSupply;

				if (this.IsRejuvenating)
					state |= FleetSituation.Rejuvenating;
			}

			var heavilyDamaged = ships
				.Where(s => !this.homeport.Repairyard.CheckRepairing(s.Id))
				.Where(s => !s.Situation.HasFlag(ShipSituation.Evacuation) && !s.Situation.HasFlag(ShipSituation.Tow))
				.Where(s => !(state.HasFlag(FleetSituation.Sortie) && s.Situation.HasFlag(ShipSituation.DamageControlled)))
				.Any(s => s.HP.IsHeavilyDamage());
			if (heavilyDamaged)
				state |= FleetSituation.HeavilyDamaged;

			if (this.Ships.Length > 0)
			{
				if (this.Ships[0].Info.ShipType.Id == 19)
					state |= FleetSituation.FlagshipIsRepairShip;
			}

			this.State = state;
		}

		private void CalculateValues()
		{
			var ships = this.Ships
				.Where(x => !x.Situation.HasFlag(ShipSituation.Tow) && !x.Situation.HasFlag(ShipSituation.Evacuation))
				.ToArray();

			this.AirSuperiorityPotentialMinimum = ships.Sum(x => x.GetAirSuperiorityPotential(AirSuperiorityCalculationOptions.Minimum));
			this.AirSuperiorityPotentialMaximum = ships.Sum(x => x.GetAirSuperiorityPotential(AirSuperiorityCalculationOptions.Maximum));

			this.LOS = LOSCalcLogic.Get(Settings.LOSCalcType).Calc(new Fleet[] { this });
		}

		protected override void Tick()
		{
			base.Tick();

			if (this.RejuvenateTime.HasValue)
			{
				this.RaisePropertyChanged(nameof(RejuvenateRemaining));
				this.RaisePropertyChanged(nameof(RejuvenateText));
			}
		}

		internal void RaiseShipsUpdated()
		{
			this.RaisePropertyChanged(nameof(this.ShipsUpdated));
		}

		public override string ToString()
		{
			return $"ID = {this.Id}, Name = \"{this.Name}\", Ships = {string.Join(",", this.Ships.Select(s => "\"" + s.Info.Name + "\""))}";
		}
	}
}

File: WindowFrame/source/System/Homeport/Fleet.ts
import { IIdentifiable } from "System/Base/Interfaces/IIdentifiable";
import { Homeport } from "./Homeport";
import { kcsapi_deck, kcsapi_req_member_updatedeckname } from "System/Interfaces/kcsapi_deck";
import { TickObservable } from "System/Base/Observable";
import { Ship } from "./Ship";
import { ShipSpeed, ShipType } from "System/Enums/ShipEnums";
import Const from "System/Const/System";
import { FleetState } from "System/Enums/FleetEnums";
import { Expedition } from "./Expedition/Expedition";
import { AirSupremacy } from "System/Models/AirSupremacy";
import { LoSCalculator } from "System/Models/LoSCalculator/LoSCalculator";
import { SubscribeKcsapi } from "System/Base/KcsApi";
import { HTTPRequest } from "System/Exports/API";
import { Settings } from "System/Settings";

export class Fleet extends TickObservable implements IIdentifiable {
    //#region Id
    private _Id: number = 0;
    public get Id(): number { return this._Id }
    //#endregion

    //#region Name
    private _Name: string = "???";
    public get Name(): string { return this._Name }
    //#endregion

    //#region Ships
    private SourceShips: (Ship | null)[] = [];
    private _Ships: Ship[] = [];
    public get Ships(): Ship[] { return this._Ships }
    //#endregion

    public readonly Expedition: Expedition;

    public get FleetSpeed(): ShipSpeed {
        return this.Ships.length > 0
            ? this.Ships.reduce((a, c) => Math.min(a, c.Speed), ShipSpeed.Fastest)
            : ShipSpeed.None;
    }

    //#region State
    private _State: FleetState = FleetState.Empty;
    public get State(): FleetState { return this._State }
    //#endregion

    //#region IsSailing
    private _IsSailing: boolean = false;
    public get IsSailing(): boolean { return this._IsSailing }
    //#endregion

    //#region ConditionRestoreTime, IsConditionRestoring
    private _ConditionRestoreTime: number = 0;
    public get ConditionRestoreTime(): number { return this._ConditionRestoreTime }
    public get IsConditionRestoring(): boolean { return this.ConditionRestoreTime > 0 }
    //#endregion

    //#region AirSupremacy
    private _AirSupremacy: AirSupremacy = new AirSupremacy();
    public get AirSupremacy(): AirSupremacy { return this._AirSupremacy }
    //#endregion

    //#region LoS
    private _LoS: number = 0;
    public get LoS(): number { return this._LoS }
    ////#endregion


    private homeport: Homeport;
    constructor(owner: Homeport, data: kcsapi_deck) {
        super();

        this.homeport = owner;
        this.ManagedDisposable.Add(this.Expedition = new Expedition(this));
        this.ManagedDisposable.Add(SubscribeKcsapi<{}, kcsapi_req_member_updatedeckname>(
            "api_req_member/updatedeckname", (x, y) => this.UpdateFleetName(y))
        );

        this.Update(data);
    }

    private UpdateFleetName(data: kcsapi_req_member_updatedeckname): void {
        const fleetId = data.api_deck_id;;
        if (this.Id !== fleetId) return;

        this.$._Name = data.api_name.toString(); // Could be numeric name like "1234"
    }

    public Update(data: kcsapi_deck): void {
        this.$._Id = data.api_id;
        this.$._Name = data.api_name;

        this.Expedition.Update(data.api_mission);
        this.UpdateShips(
            (<Ship[]>data.api_ship.map(x => this.homeport.Ships.get(x)).filter(x => x))
        );

        this.UpdateCondition();
    }

    private UpdateShips(ships: Ship[]): void {
        this.SourceShips = ships;
        this.$._Ships = ships.filter(x => x !== null);

        this.Calculate();
        this.UpdateState();
        this.ShipUpdated();
    }

    private PrevCondition!: number;
    private UpdateCondition(): void {
        const condition: number = this.Ships.reduce((a, c) => a === 0 ? c.Condition : Math.min(a, c.Condition), 0);
        const goal = Const.ConditionRestorationLimit;

        // Require time not changed
        if (this.PrevCondition === condition) return;

        this.PrevCondition = condition;
        if (condition >= goal)
            this.$._ConditionRestoreTime = 0;
        else {
            let restore = Date.now();

            const value = Math.floor((goal - condition + 2) / 3) * 3; // Integer dividing
            restore = restore + value * 60 * 1000;

            this.$._ConditionRestoreTime = restore;
        }
        this.RaisePropertyChanged(nameof(this.IsConditionRestoring));
        this.UpdateState();
    }

    public UpdateState(): void {
        let state = FleetState.Empty;

        const ships = this.Ships;
        if (ships.length > 0) {
            if (this.IsSailing)
                state |= FleetState.Sailing;
            else if (this.Expedition.IsInExecution)
                state |= FleetState.Expedition;
            else
                state |= FleetState.Homeport;
        }

        if ((state & FleetState.Homeport) !== 0) {
            // Check repairing
            if (ships.some(x => this.homeport.RepairDock!.CheckRepairing(x.Id)))
                state |= FleetState.Repairing;

            // Not fully supplied
            if (ships.some(x => x.Fuel.Current < x.Fuel.Maximum || x.Ammo.Current < x.Ammo.Maximum))
                state |= FleetState.NeedSupply;

            // Condition restoring
            if (this.IsConditionRestoring)
                state |= FleetState.ConditionRestoring;
        }

        const heavilyDamaged: boolean = ships
            .filter(x => !this.homeport.RepairDock!.CheckRepairing(x.Id))
            .filter(x => !(x.State & Ship.State.Evacuation) && !(x.State & Ship.State.Tow))
            .filter(x => !(state & FleetState.Sailing) && (x.State & Ship.State.DamageControlled))
            .some(x => x.HP.Percentage <= 0.25);
        if (heavilyDamaged)
            state |= FleetState.HeavilyDamaged;

        if (ships.length > 0) {
            if (ships[0].Info.ShipType && (ships[0].Info.ShipType.Id == ShipType.RepairShip))
                state |= FleetState.FlagshipRepairShip;
        }

        this.$._State = state;
    }
    public Calculate(): void {
        const ships = this.Ships.filter(x => // Only actually exists in fleet
            (x.State & Ship.State.Tow) === 0
            && (x.State & Ship.State.Evacuation) === 0
        );

        this.$._AirSupremacy = AirSupremacy.Sum(ships.map(x => x.AirSupremacy));

        const calculator = LoSCalculator.Instance.Get(Settings.Instance.LoSCalculator);
        if (calculator)
            this.$._LoS = calculator.Calc([this]);
        else
            this.$._LoS = 0;
    }

    public UnsetAll(): void {
        const list = [this.SourceShips[0]];
        const ships = new Array(this.SourceShips.length);
        for (let i = 0; i < list.length; i++)
            ships[i] = list[i];

        this.UpdateShips(ships);
    }
    public Unset(index: number): void {
        const list = this.SourceShips.filter((_, y) => y !== index);
        const ships = new Array(this.SourceShips.length);
        for (let i = 0; i < list.length; i++)
            ships[i] = list[i];

        this.UpdateShips(ships);
    }
    public Change(index: number, ship: Ship): Ship {
        const current = this.SourceShips[index];

        let list: Ship[] = [];
        if (index == -1) {
            list = [(<Ship>this.SourceShips.find(x => true))];
        }
        else {
            const temp = this.SourceShips;
            temp[index] = ship;
            list = <Ship[]>temp.filter(x => x);
        }

        const ships = new Array(this.SourceShips.length);
        for (let i = 0; i < list.length; i++)
            ships[i] = list[i];

        this.UpdateShips(ships);

        return <Ship>current;
    }
    public ShipUpdated(): void {
        this.RaisePropertyChanged(nameof(this.Ships));
        this.RaisePropertyChanged(nameof(this.FleetSpeed));
    }

    protected Tick(): void {
        super.Tick();

        if(this.IsConditionRestoring){
            this.RaisePropertyChanged(nameof(this.ConditionRestoreTime));
        }
    }
}